### ISSUE TITLE
TLS handshake errors in logs due to bosh health check 

### ISSUE THREAD

**Issue by axel7born** [cloudfoundry/nats-release#32]
We observed TLS handshake errors in nats-tls logs:
```
[7] 2020/11/18 14:21:38.776820 [ERR] 10.0.129.3:47892 - cid:12 - TLS handshake error: EOF
[7] 2020/11/18 14:21:48.824900 [ERR] 10.0.129.3:47908 - cid:13 - TLS handshake error: EOF
[7] 2020/11/18 14:21:58.829983 [ERR] 10.0.129.3:47950 - cid:14 - TLS handshake error: EOF
[7] 2020/11/18 14:22:08.835211 [ERR] 10.0.129.3:47970 - cid:15 - TLS handshake error: EOF
[7] 2020/11/18 14:22:18.840149 [ERR] 10.0.129.3:48020 - cid:16 - TLS handshake error: EOF
[7] 2020/11/18 14:22:28.845073 [ERR] 10.0.129.3:48034 - cid:17 - TLS handshake error: EOF
[7] 2020/11/18 14:22:38.850086 [ERR] 10.0.129.3:48078 - cid:18 - TLS handshake error: EOF
[7] 2020/11/18 14:22:48.856597 [ERR] 10.0.129.3:48092 - cid:19 - TLS handshake error: EOF
[7] 2020/11/18 14:22:58.861785 [ERR] 10.0.129.3:48134 - cid:20 - TLS handshake error: EOF
[7] 2020/11/18 14:23:08.866729 [ERR] 10.0.129.3:48150 - cid:21 - TLS handshake error: EOF
[7] 2020/11/18 14:23:18.871733 [ERR] 10.0.129.3:48204 - cid:22 - TLS handshake error: EOF
[7] 2020/11/18 14:23:28.876737 [ERR] 10.0.129.3:48218 - cid:23 - TLS handshake error: EOF
[7] 2020/11/18 14:23:38.881706 [ERR] 10.0.129.3:48260 - cid:24 - TLS handshake error: EOF
[7] 2020/11/18 14:23:48.886635 [ERR] 10.0.129.3:48274 - cid:25 - TLS handshake error: EOF
[7] 2020/11/18 14:23:58.891747 [ERR] 10.0.129.3:48312 - cid:26 - TLS handshake error: EOF
```
The reason for the failed handshake is the bosh health check introduced with 140e2e862b4302aaa211eaa781e811601f9c0f48

The log entries are confusing and it took us quite some to find out what the cause is.

**Comment by ameowlia** [comment]
Hi @axel7born can you please provide more information. For example, some helpful bits information might be:

❓ What debugging have you done so far to determine why the health check is failing?
❓ When did you see this error? Immediately after a deployment or had the deployment been successfully running and then started failing? 
❓ Did this happen on all nats VMs? 
❓ Were your nats-tls processes healthy? What did the CPU stats look like at the time? 
❓ Are you able to manually connect to the nats-tls process? 
❓ How can I reproduce this?

**Comment by axel7born** [comment]
Hi @ameowlia,
The health check is not failing.
The error logs are the normal behaviour since the introduction of the monit health check.
The tcp check of monit opens a tcp connection but doesn't do the handshake.
With tcpdump, we can see that monit sends a syn and after receiving an ack it sends a fin
```
tcpdump -n -i lo port 4224
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on lo, link-type EN10MB (Ethernet), capture size 262144 bytes
07:51:44.241518 IP 10.0.129.3.57838 > 10.0.129.3.4224: Flags [S], seq 604788005, win 65495, options [mss 65495,sackOK,TS val 4045862511 ecr 0,nop,wscale 7], length 0
07:51:44.241524 IP 10.0.129.3.4224 > 10.0.129.3.57838: Flags [S.], seq 2802588045, ack 604788006, win 65483, options [mss 65495,sackOK,TS val 4045862511 ecr 4045862511,nop,wscale 7], length 0
07:51:44.241529 IP 10.0.129.3.57838 > 10.0.129.3.4224: Flags [.], ack 1, win 512, options [nop,nop,TS val 4045862511 ecr 4045862511], length 0
07:51:44.241551 IP 10.0.129.3.57838 > 10.0.129.3.4224: Flags [F.], seq 1, ack 1, win 512, options [nop,nop,TS val 4045862511 ecr 4045862511], length 0
07:51:44.241707 IP 10.0.129.3.4224 > 10.0.129.3.57838: Flags [P.], seq 1:221, ack 2, win 512, options [nop,nop,TS val 4045862511 ecr 4045862511], length 220
07:51:44.241721 IP 10.0.129.3.57838 > 10.0.129.3.4224: Flags [R], seq 604788007, win 0, length 0
07:51:54.246294 IP 10.0.129.3.57852 > 10.0.129.3.4224: Flags [S], seq 1101576446, win 65495, options [mss 65495,sackOK,TS val 4045872516 ecr 0,nop,wscale 7], length 0
07:51:54.246302 IP 10.0.129.3.4224 > 10.0.129.3.57852: Flags [S.], seq 1352843357, ack 1101576447, win 65483, options [mss 65495,sackOK,TS val 4045872516 ecr 4045872516,nop,wscale 7], length 0
07:51:54.246308 IP 10.0.129.3.57852 > 10.0.129.3.4224: Flags [.], ack 1, win 512, options [nop,nop,TS val 4045872516 ecr 4045872516], length 0
07:51:54.246337 IP 10.0.129.3.57852 > 10.0.129.3.4224: Flags [F.], seq 1, ack 1, win 512, options [nop,nop,TS val 4045872516 ecr 4045872516], length 0
07:51:54.246474 IP 10.0.129.3.4224 > 10.0.129.3.57852: Flags [P.], seq 1:221, ack 2, win 512, options [nop,nop,TS val 4045872516 ecr 4045872516], length 220
07:51:54.246485 IP 10.0.129.3.57852 > 10.0.129.3.4224: Flags [R], seq 1101576448, win 0, length 0
```
This results in a failing handshake  reported here:
https://github.com/nats-io/nats-server/blob/16d7d032e14682a9dccaf7c26f7eda285639d021/server/server.go#L2237

That happens to nats-tls processes on all VMs. For the nats process, the problem doesn't exist.
  
To reproduce, just look at the  output of  `tail -f /var/vcap/sys/log/nats-tls/nats-tls.stderr.log`
You should see a new error log entry each 10s.

**Comment by wallyqs** [comment]
Wouldn't it be better to have monit make a tcp check against port 8222? There is no `http://localhost/healthz` but sending a request to `http://localhost:8222/` or `http://localhost:8222/varz` would serve the same.

**Comment by ameowlia** [comment]
Thank you for the further explanation @axel7born, I understand now. And thank you for that suggestion @wallyqs. 

I will put a story to fix this in our backlog, but a PR would be much appreciated and are often the fastest way to fix an issue 😄

**Comment by geofffranks** [comment]
👋 hey @axel7born - sorry it's been so long with no updates on this. There's currently some discussion going on in PR #37 to provide a possible solution for you. If you have a minute to take a look and see if it'll meet your needs, I'd appreciate it

**Comment by geofffranks** [comment]
This should be fixed now by #42 , which we'll release shortly after it runs through pipelines
